update to go1.22.6

thaJeztah opened the pull request:
- [x] depends on https://github.com/moby/moby/pull/47016
- [x] depends on https://github.com/microsoft/go-winio/pull/312
- [x] depends on https://github.com/moby/moby/pull/47743
- [x] depends on https://github.com/moby/moby/pull/47977
- [x] depends on https://github.com/moby/moby/pull/47976 
- [x] depends on https://github.com/moby/moby/pull/48359


**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog.
It must be placed inside the below triple backticks section:
-->
```markdown changelog
update to go1.22.6
```

**- A picture of a cute animal (not mandatory but encouraged)**

thaJeztah committed:
update to go1.22.6

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented:
Arf.. here we go again with pre-releases and the odd version format;

```
Setup go version spec 1.22rc1
Attempting to download 1.22rc1...
matching 1.22rc1...
Not found in manifest.  Falling back to download directly from Go
Error: Unable to find Go version '1.22rc1' for platform linux and architecture x64.
```

thaJeztah commented:
Interesting; two failures: a panic in golang-ci-lint;

```
ERRO [runner] Panic: cgocall: package "sdjournal" (isInitialPkg: true, needAnalyzeSource: true): runtime error: invalid memory address or nil pointer dereference: goroutine 28078 [running]:
runtime/debug.Stack()
	/usr/local/go/src/runtime/debug/stack.go:24 +0x5e
github.com/golangci/golangci-lint/pkg/golinters/goanalysis.(*action).analyzeSafe.func1()
	/go/pkg/mod/github.com/golangci/golangci-lint@v1.54.2/pkg/golinters/goanalysis/runner_action.go:109 +0x277
panic({0x1647820?, 0x222b380?})
	/usr/local/go/src/runtime/panic.go:770 +0x132
go/types.(*Checker).handleBailout(0xc01126f200, 0xc0115d7a48)
	/usr/local/go/src/go/types/check.go:367 +0x88
panic({0x1647820?, 0x222b380?})
	/usr/local/go/src/runtime/panic.go:770 +0x132
go/types.(*StdSizes).Sizeof(0x0, {0x1a39980, 0x2243120})
	/usr/local/go/src/go/types/sizes.go:228 +0x31e
go/types.(*Config).sizeof(...)
	/usr/local/go/src/go/types/sizes.go:333
go/types.representableConst.func1({0x1a39980?, 0x2243120?})
	/usr/local/go/src/go/types/const.go:76 +0x9e
go/types.representableConst({0x1a3f448, 0x23844e0}, 0xc01126f200, 0x2243120, 0xc0115d4e80)
	/usr/local/go/src/go/types/const.go:92 +0x192
go/types.(*Checker).representation(0xc01126f200, 0xc0115dab40, 0x2243120)
	/usr/local/go/src/go/types/const.go:256 +0x65
go/types.(*Checker).implicitTypeAndValue(0xc01126f200, 0xc0115dab40, {0x1a399d0, 0xc00d21ef50})
	/usr/local/go/src/go/types/expr.go:375 +0x30d
go/types.(*Checker).assignment(0xc01126f200, 0xc0115dab40, {0x1a399d0, 0xc00d21ef50}, {0x181de26, 0x10})
	/usr/local/go/src/go/types/assignments.go:52 +0x2e5
go/types.(*Checker).initVar(0xc01126f200, 0xc0115b7500, 0xc0115dab40, {0x181de26, 0x10})
	/usr/local/go/src/go/types/assignments.go:163 +0x41e
go/types.(*Checker).initVars(0xc01126f200, {0xc011543d60, 0x2, 0x1a39980?}, {0xc011579800, 0x22430e0?, 0xc0115c15c0?}, {0x1a3d7f8, 0xc011579820})
	/usr/local/go/src/go/types/assignments.go:383 +0x638
go/types.(*Checker).stmt(0xc01126f200, 0x0, {0x1a3d7f8, 0xc011579820})
	/usr/local/go/src/go/types/stmt.go:524 +0x1fb7
go/types.(*Checker).stmtList(0xc01126f200, 0x0, {0xc011543010?, 0x17a002a?, 0x5?})
	/usr/local/go/src/go/types/stmt.go:121 +0x85
go/types.(*Checker).stmt(0xc01126f200, 0x0, {0x1a3d888, 0xc011576b10})
	/usr/local/go/src/go/types/stmt.go:562 +0x20e5
go/types.(*Checker).stmt(0xc01126f200, 0x0, {0x1a3d8b8, 0xc01157c280})
	/usr/local/go/src/go/types/stmt.go:574 +0x2ff7
go/types.(*Checker).stmtList(0xc01126f200, 0x0, {0xc01157c300?, 0x0?, 0x0?})
	/usr/local/go/src/go/types/stmt.go:121 +0x85
go/types.(*Checker).funcBody(0xc01126f200, 0x1a399d0?, {0x1a2b6d8?, 0xc000154000?}, 0xc0115a6e00, 0xc011576b40, {0x0?, 0x0?})
	/usr/local/go/src/go/types/stmt.go:41 +0x331
go/types.(*Checker).funcDecl.func1()
	/usr/local/go/src/go/types/decl.go:852 +0x3a
go/types.(*Checker).processDelayed(0xc01126f200, 0x0)
	/usr/local/go/src/go/types/check.go:467 +0x162
go/types.(*Checker).checkFiles(0xc01126f200, {0xc01152a4e8, 0x1, 0x1})
	/usr/local/go/src/go/types/check.go:411 +0x1cc
go/types.(*Checker).Files(...)
	/usr/local/go/src/go/types/check.go:372
go/types.(*Config).Check(0xc010bd76e0, {0xc001131950?, 0x5b?}, 0xc002085b40, {0xc01152a4e8, 0x1, 0x1}, 0xc010bd7740)
	/usr/local/go/src/go/types/api.go:458 +0x79
golang.org/x/tools/go/analysis/passes/cgocall.typeCheckCgoSourceFiles(0xc002085b40, 0xc00c93ce40, {0xc00d3072a0, 0x4, 0x4?}, 0xc00c93cea0, {0x1a3cdd8, 0x0})
	/go/pkg/mod/golang.org/x/tools@v0.12.0/go/analysis/passes/cgocall/cgocall.go:281 +0x3fc
golang.org/x/tools/go/analysis/passes/cgocall.run(0xc0111524e0)
	/go/pkg/mod/golang.org/x/tools@v0.12.0/go/analysis/passes/cgocall/cgocall.go:48 +0xb7
github.com/golangci/golangci-lint/pkg/golinters/goanalysis.(*action).analyze(0xc004dee080)
	/go/pkg/mod/github.com/golangci/golangci-lint@v1.54.2/pkg/golinters/goanalysis/runner_action.go:195 +0x99a
github.com/golangci/golangci-lint/pkg/golinters/goanalysis.(*action).analyzeSafe.func2()
	/go/pkg/mod/github.com/golangci/golangci-lint@v1.54.2/pkg/golinters/goanalysis/runner_action.go:113 +0x17
github.com/golangci/golangci-lint/pkg/timeutils.(*Stopwatch).TrackStage(0xc0016b0b90, {0x17a5ea9, 0x7}, 0xc00690ff48)
	/go/pkg/mod/github.com/golangci/golangci-lint@v1.54.2/pkg/timeutils/stopwatch.go:111 +0x44
github.com/golangci/golangci-lint/pkg/golinters/goanalysis.(*action).analyzeSafe(0xc002c89c20?)
	/go/pkg/mod/github.com/golangci/golangci-lint@v1.54.2/pkg/golinters/goanalysis/runner_action.go:112 +0x7a
github.com/golangci/golangci-lint/pkg/golinters/goanalysis.(*loadingPackage).analyze.func2(0xc004dee080)
	/go/pkg/mod/github.com/golangci/golangci-lint@v1.54.2/pkg/golinters/goanalysis/runner_loadingpackage.go:80 +0xa8
created by github.com/golangci/golangci-lint/pkg/golinters/goanalysis.(*loadingPackage).analyze in goroutine 958
	/go/pkg/mod/github.com/golangci/golangci-lint@v1.54.2/pkg/golinters/goanalysis/runner_loadingpackage.go:75 +0x205 
```

One in cross/linux/armv5 (so 32-bit cross-compile);

```
#55 [build 6/6] RUN --mount=type=bind,target=.,rw     --mount=type=tmpfs,target=cli/winresources/dockerd     --mount=type=tmpfs,target=cli/winresources/docker-proxy     --mount=type=cache,target=/root/.cache/go-build,id=moby-build-linux/arm/v5 <<EOT (set -e...)
#55 72.09 + rm -f /go/src/github.com/docker/docker/go.mod
#55 ERROR: process "/bin/sh -c   set -e\n  target=$([ \"$DOCKER_STATIC\" = \"1\" ] && echo \"binary\" || echo \"dynbinary\")\n  xx-go --wrap\n  PKG_CONFIG=$(xx-go env PKG_CONFIG) ./hack/make.sh $target\n  xx-verify $([ \"$DOCKER_STATIC\" = \"1\" ] && echo \"--static\") /tmp/bundles/${target}-daemon/dockerd$([ \"$(xx-info os)\" = \"windows\" ] && echo \".exe\")\n  xx-verify $([ \"$DOCKER_STATIC\" = \"1\" ] && echo \"--static\") /tmp/bundles/${target}-daemon/docker-proxy$([ \"$(xx-info os)\" = \"windows\" ] && echo \".exe\")\n  mkdir /build\n  mv /tmp/bundles/${target}-daemon/* /build/\n" did not complete successfully: exit code: 1
------
 > [build 6/6] RUN --mount=type=bind,target=.,rw     --mount=type=tmpfs,target=cli/winresources/dockerd     --mount=type=tmpfs,target=cli/winresources/docker-proxy     --mount=type=cache,target=/root/.cache/go-build,id=moby-build-linux/arm/v5 <<EOT (set -e...):
0.796 + trap 'rm -f "${ROOTDIR}/go.mod"' EXIT
0.796 + GO111MODULE=on
0.796 + go build -mod=vendor -modfile=vendor.mod -o /tmp/bundles/binary-daemon/dockerd -tags 'netgo osusergo static_build  journald' -ldflags '-w -X "github.com/docker/docker/dockerversion.Version=dev" -X "github.com/docker/docker/dockerversion.GitCommit=HEAD" -X "github.com/docker/docker/dockerversion.BuildTime=2024-01-03T12:18:00.000000000+00:00" -X "github.com/docker/docker/dockerversion.PlatformName=" -X "github.com/docker/docker/dockerversion.ProductName=" -X "github.com/docker/docker/dockerversion.DefaultProductLicense="  -extldflags -static ' -gcflags= github.com/docker/docker/cmd/dockerd
5.968 # runtime/cgo
5.968 gcc_libinit.c:44:8: error: large atomic operation may incur significant performance penalty; the access size (4 bytes) exceeds the max lock-free size (0  bytes) [-Werror,-Watomic-alignment]
5.968 gcc_libinit.c:47:6: error: large atomic operation may incur significant performance penalty; the access size (4 bytes) exceeds the max lock-free size (0  bytes) [-Werror,-Watomic-alignment]
5.968 gcc_libinit.c:49:10: error: large atomic operation may incur significant performance penalty; the access size (4 bytes) exceeds the max lock-free size (0  bytes) [-Werror,-Watomic-alignment]
5.968 gcc_libinit.c:69:9: error: large atomic operation may incur significant performance penalty; the access size (4 bytes) exceeds the max lock-free size (0  bytes) [-Werror,-Watomic-alignment]
5.968 gcc_libinit.c:71:3: error: large atomic operation may incur significant performance penalty; the access size (4 bytes) exceeds the max lock-free size (0  bytes) [-Werror,-Watomic-alignment]
72.09 + rm -f /go/src/github.com/docker/docker/go.mod
```

thaJeztah commented:
Ah; the good news; golangci-lint panic seems to be fixed.

The bad news; looks like Windows is pretty much broken;

```
failed to register layer: re-exec error: exit status 1: output: SetFileInformationByHandle \\?\C:\Windows\SystemTemp\hcs2001018820\Files.$wcidirs$: Invalid access to memory location.
Error: Process completed with exit code 1.
```

thaJeztah commented:
Friends at Microsoft are looking into this; tracking issue on their Go fork is here;

- https://github.com/microsoft/go/issues/1100

thaJeztah commented:
Looks like it's am in go-winio after all, and an alignment issue;

- go issue https://github.com/golang/go/issues/65069
- fix in go-winio: https://github.com/microsoft/go-winio/pull/312

dagood commented:
The go-winio fix is merged: https://github.com/microsoft/go-winio/pull/312, `github.com/Microsoft/go-winio@008bc6ea43`

thaJeztah commented:
Thanks! I rebased, and updated go-winio to current master to see if CI is happier 🤞

thaJeztah commented:
CI looks a lot happier now!


The only failure is cross-compile for `linux/arm/v5`. But looking at some output... interestingly it also shows alignment issues?


```
> [build 6/6] RUN --mount=type=bind,target=.,rw     --mount=type=tmpfs,target=cli/winresources/dockerd     --mount=type=tmpfs,target=cli/winresources/docker-proxy     --mount=type=cache,target=/root/.cache/go-build,id=moby-build-linux/arm/v5 <<EOT (set -e...):
0.786 + trap 'rm -f "${ROOTDIR}/go.mod"' EXIT
0.786 + GO111MODULE=on
0.786 + go build -mod=vendor -modfile=vendor.mod -o /tmp/bundles/binary-daemon/dockerd -tags 'netgo osusergo static_build  journald' -ldflags '-w -X "github.com/docker/docker/dockerversion.Version=dev" -X "github.com/docker/docker/dockerversion.GitCommit=HEAD" -X "github.com/docker/docker/dockerversion.BuildTime=2024-01-17T13:52:01.000000000+00:00" -X "github.com/docker/docker/dockerversion.PlatformName=" -X "github.com/docker/docker/dockerversion.ProductName=" -X "github.com/docker/docker/dockerversion.DefaultProductLicense="  -extldflags -static ' -gcflags= github.com/docker/docker/cmd/dockerd
7.003 # runtime/cgo
7.003 gcc_libinit.c:44:8: error: large atomic operation may incur significant performance penalty; the access size (4 bytes) exceeds the max lock-free size (0  bytes) [-Werror,-Watomic-alignment]
7.003 gcc_libinit.c:47:6: error: large atomic operation may incur significant performance penalty; the access size (4 bytes) exceeds the max lock-free size (0  bytes) [-Werror,-Watomic-alignment]
7.003 gcc_libinit.c:49:10: error: large atomic operation may incur significant performance penalty; the access size (4 bytes) exceeds the max lock-free size (0  bytes) [-Werror,-Watomic-alignment]
7.003 gcc_libinit.c:69:9: error: large atomic operation may incur significant performance penalty; the access size (4 bytes) exceeds the max lock-free size (0  bytes) [-Werror,-Watomic-alignment]
7.003 gcc_libinit.c:71:3: error: large atomic operation may incur significant performance penalty; the access size (4 bytes) exceeds the max lock-free size (0  bytes) [-Werror,-Watomic-alignment]
74.43 + rm -f /go/src/github.com/docker/docker/go.mod
------
Dockerfile:610
--------------------
 609 |     EOT
 610 | >>> RUN --mount=type=bind,target=.,rw \
 611 | >>>     --mount=type=tmpfs,target=cli/winresources/dockerd \
 612 | >>>     --mount=type=tmpfs,target=cli/winresources/docker-proxy \
 613 | >>>     --mount=type=cache,target=/root/.cache/go-build,id=moby-build-$TARGETPLATFORM <<EOT
 614 | >>>   set -e
 615 | >>>   target=$([ "$DOCKER_STATIC" = "1" ] && echo "binary" || echo "dynbinary")
 616 | >>>   xx-go --wrap
 617 | >>>   PKG_CONFIG=$(xx-go env PKG_CONFIG) ./hack/make.sh $target
 618 | >>>   xx-verify $([ "$DOCKER_STATIC" = "1" ] && echo "--static") /tmp/bundles/${target}-daemon/dockerd$([ "$(xx-info os)" = "windows" ] && echo ".exe")
 619 | >>>   xx-verify $([ "$DOCKER_STATIC" = "1" ] && echo "--static") /tmp/bundles/${target}-daemon/docker-proxy$([ "$(xx-info os)" = "windows" ] && echo ".exe")
 620 | >>>   mkdir /build
 621 | >>>   mv /tmp/bundles/${target}-daemon/* /build/
 622 | >>> EOT
 623 |     
--------------------
ERROR: failed to solve: process "/bin/sh -c   set -e\n  target=$([ \"$DOCKER_STATIC\" = \"1\" ] && echo \"binary\" || echo \"dynbinary\")\n  xx-go --wrap\n  PKG_CONFIG=$(xx-go env PKG_CONFIG) ./hack/make.sh $target\n  xx-verify $([ \"$DOCKER_STATIC\" = \"1\" ] && echo \"--static\") /tmp/bundles/${target}-daemon/dockerd$([ \"$(xx-info os)\" = \"windows\" ] && echo \".exe\")\n  xx-verify $([ \"$DOCKER_STATIC\" = \"1\" ] && echo \"--static\") /tmp/bundles/${target}-daemon/docker-proxy$([ \"$(xx-info os)\" = \"windows\" ] && echo \".exe\")\n  mkdir /build\n  mv /tmp/bundles/${target}-daemon/* /build/\n" did not complete successfully: exit code: 1
Error: buildx bake failed with: ERROR: failed to solve: process "/bin/sh -c   set -e\n  target=$([ \"$DOCKER_STATIC\" = \"1\" ] && echo \"binary\" || echo \"dynbinary\")\n  xx-go --wrap\n  PKG_CONFIG=$(xx-go env PKG_CONFIG) ./hack/make.sh $target\n  xx-verify $([ \"$DOCKER_STATIC\" = \"1\" ] && echo \"--static\") /tmp/bundles/${target}-daemon/dockerd$([ \"$(xx-info os)\" = \"windows\" ] && echo \".exe\")\n  xx-verify $([ \"$DOCKER_STATIC\" = \"1\" ] && echo \"--static\") /tmp/bundles/${target}-daemon/docker-proxy$([ \"$(xx-info os)\" = \"windows\" ] && echo \".exe\")\n  mkdir /build\n  mv /tmp/bundles/${target}-daemon/* /build/\n" did not complete successfully: exit code: 1
```

That appears _seems_ to come from CLANG; https://github.com/c2lang/c2compiler/blob/82a0cc5a79860cba8555cd6ee95d278f34a8ecce/c2c/Clang/DiagnosticSemaKinds.inc#L2764 And that message also mentions "alignment", so now wondering if things are related to the issue we saw on Windows :thinking_face:

```cgo
DIAG(warn_atomic_op_misaligned, CLASS_WARNING, (unsigned)diag::Severity::Warning, "misaligned or large atomic operation may incur significant performance penalty", 41, SFINAE_Suppress, false, false, 2)
```

A quick search also gave me some code-bases having exceptions for this; https://github.com/kokkos/kokkos/blob/71a9bcae52543bd065522bf3e41b5bfa467d8015/tpls/desul/include/desul/atomics/Compare_Exchange_GCC.hpp#L33-L41

```cgo
// clang-format off
// Disable warning for large atomics on clang 7 and up (checked with godbolt)
// error: large atomic operation may incur significant performance penalty [-Werror,-Watomic-alignment]
// https://godbolt.org/z/G7YhqhbG6
// clang-format on
#if defined(__clang__) && (__clang_major__ >= 7) && !defined(__APPLE__)
#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Watomic-alignment"
#endif
```

https://github.com/AaronRobinsonMSFT/DNNE/blob/3e39b7ca22dea40c4346be7e94810248ab402b38/src/platform/platform.c#L291-L297

```cgo
#ifdef __clang__
#pragma clang diagnostic push
#ifdef __arm__
// warning: large atomic operation may incur significant performance penalty; the access size (4 bytes) exceeds the max lock-free size (0  bytes)
#pragma clang diagnostic ignored "-Watomic-alignment"
#endif // __arm__
#endif // __clang__
```

dagood commented:
> gcc_libinit.c:44

I checked for changes from go1.21.5 -> go1.22rc1, and all the lines with these warnings are actually new to 1.22: https://github.com/golang/go/commit/b104a0ef32e1388dbd6af7174975857320fb8e80, https://github.com/golang/go/issues/60961.

I don't know off the top of my head if ignoring the warnings is something that can be done in the code here, or whether that would mask some actually-bad effect here. Seems worth raising upstream.

thaJeztah commented:
Thanks Cory!

I'll do a rebase of this one, and update to rc2 as well.

thaJeztah commented:
updated to go1.22.1

we don't have a tagged go-winio release with the patch yet though

thaJeztah commented:
arm/v5 is still failing I see; something to look into 🤔

thaJeztah commented:
```
> [build 6/6] RUN --mount=type=bind,target=.,rw     --mount=type=tmpfs,target=cli/winresources/dockerd     --mount=type=tmpfs,target=cli/winresources/docker-proxy     --mount=type=cache,target=/root/.cache/go-build,id=moby-build-linux/arm/v5 <<EOT (set -e...):
0.821 + GO111MODULE=on
0.821 + GOTOOLCHAIN=local
0.822 + go build -mod=vendor -modfile=vendor.mod -o /tmp/bundles/binary-daemon/dockerd -tags 'netgo osusergo static_build  journald' -ldflags '-w -X "github.com/docker/docker/dockerversion.Version=dev" -X "github.com/docker/docker/dockerversion.GitCommit=HEAD" -X "github.com/docker/docker/dockerversion.BuildTime=2024-06-13T16:12:01.000000000+00:00" -X "github.com/docker/docker/dockerversion.PlatformName=" -X "github.com/docker/docker/dockerversion.ProductName=" -X "github.com/docker/docker/dockerversion.DefaultProductLicense="  -extldflags -static ' -gcflags= github.com/docker/docker/cmd/dockerd
6.745 # runtime/cgo
6.745 gcc_libinit.c:44:8: error: large atomic operation may incur significant performance penalty; the access size (4 bytes) exceeds the max lock-free size (0  bytes) [-Werror,-Watomic-alignment]
6.745 gcc_libinit.c:47:6: error: large atomic operation may incur significant performance penalty; the access size (4 bytes) exceeds the max lock-free size (0  bytes) [-Werror,-Watomic-alignment]
6.745 gcc_libinit.c:49:10: error: large atomic operation may incur significant performance penalty; the access size (4 bytes) exceeds the max lock-free size (0  bytes) [-Werror,-Watomic-alignment]
6.745 gcc_libinit.c:69:9: error: large atomic operation may incur significant performance penalty; the access size (4 bytes) exceeds the max lock-free size (0  bytes) [-Werror,-Watomic-alignment]
6.745 gcc_libinit.c:71:3: error: large atomic operation may incur significant performance penalty; the access size (4 bytes) exceeds the max lock-free size (0  bytes) [-Werror,-Watomic-alignment]
74.94 + rm -f /go/src/github.com/docker/docker/go.mod
```

thaJeztah commented:
Fix for the above will be in go1.23; https://github.com/golang/go/blob/956f8a67dd7319bad60c015d982f0b2e95b9f382/src/runtime/cgo/gcc_libinit.c#L7-L12

thaJeztah commented:
Added a commit to update to go1.23rc1 to see if it indeed fixes the issue.

thaJeztah commented:
arm/v5 still failing, but looks different now;

```
#56 84.12 /usr/bin/arm-linux-gnueabi-clang -marm -o $WORK/b001/exe/a.out -Wl,--export-dynamic-symbol=_cgo_panic -Wl,--export-dynamic-symbol=_cgo_topofstack -Wl,--export-dynamic-symbol=crosscall2 -Qunused-arguments -Wl,--compress-debug-sections=zlib /tmp/go-link-759578347/go.o /tmp/go-link-759578347/000000.o /tmp/go-link-759578347/000001.o /tmp/go-link-759578347/000002.o /tmp/go-link-759578347/000003.o /tmp/go-link-759578347/000004.o /tmp/go-link-759578347/000005.o /tmp/go-link-759578347/000006.o /tmp/go-link-759578347/000007.o /tmp/go-link-759578347/000008.o /tmp/go-link-759578347/000009.o /tmp/go-link-759578347/000010.o /tmp/go-link-759578347/000011.o /tmp/go-link-759578347/000012.o /tmp/go-link-759578347/000013.o /tmp/go-link-759578347/000014.o /tmp/go-link-759578347/000015.o /tmp/go-link-759578347/000016.o /tmp/go-link-759578347/000017.o /tmp/go-link-759578347/000018.o -O2 -g -O2 -g -O2 -g -lpthread -O2 -g -no-pie -static
#56 84.12 ld.lld: error: undefined symbol: __atomic_load_4
#56 84.12 >>> referenced by gcc_libinit.c
#56 84.12 >>>               /tmp/go-link-759578347/000009.o:(_cgo_wait_runtime_init_done)
#56 84.12 >>> referenced by gcc_libinit.c
#56 84.12 >>>               /tmp/go-link-759578347/000009.o:(_cgo_wait_runtime_init_done)
#56 84.12 >>> referenced by gcc_libinit.c
#56 84.12 >>>               /tmp/go-link-759578347/000009.o:(_cgo_wait_runtime_init_done)
#56 84.12 >>> referenced 2 more times
#56 84.12 
#56 84.12 ld.lld: error: undefined symbol: __atomic_store_4
#56 84.12 >>> referenced by gcc_libinit.c
#56 84.12 >>>               /tmp/go-link-759578347/000009.o:(_cgo_wait_runtime_init_done)
#56 84.12 >>> referenced by gcc_libinit.c
#56 84.12 >>>               /tmp/go-link-759578347/000009.o:(x_cgo_notify_runtime_init_done)
#56 84.12 >>> referenced by gcc_libinit.c
#56 84.12 >>>               /tmp/go-link-759578347/000009.o:(x_cgo_set_context_function)
#56 84.12 clang: error: linker command failed with exit code 1 (use -v to see invocation)

```

GolangCI-lint currently doesn't work, so must have an update as well;

```
ERRO [linters_context/goanalysis] buildir: panic during analysis: Cannot range over: func(yield func(K, V) bool), goroutine 24271 [running]:
ERRO [linters_context/goanalysis] typedness: panic during analysis: interface conversion: interface {} is nil, not *buildir.IR, goroutine 24326 [running]:
```

corhere commented:
The arm/v5 build failure looks to be due to libatomic not being included in the link. For reasons probably buried in mailing list archives, [gcc](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=81358) and clang don't bother to implicitly auto-link libatomic. This is not a big deal on many modern platforms with atomic intrinsics as the compiler generates inline instruction sequences, avoiding any libcalls into libatomic. ARMv5 is not one of those platforms: all atomic operations require a libcall. In theory, adding `CGO_LDFLAGS=-latomic` should fix arm/v5 builds.

While it could be argued that cgo should automatically link against libatomic in the same way that it automatically links against libpthread, the Go maintainers would have a valid counter-argument that it should be the C toolchain's responsibility to link against libatomic automatically, just like it does with libgcc or compiler-rt.

thaJeztah committed:
update golangci-lint to v1.60.2

Update to add go1.23 support

full diff: https://github.com/golangci/golangci-lint/compare/v1.59.1...v1.60.2
Changelog: https://golangci-lint.run/product/changelog/#1602

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
golangci-lint: temporarily disable G115: integer overflow conversion

it produces many hits, some of which may be false positives, but we need to
look into these, e.g.;

    container/container.go:517:72: G115: integer overflow conversion int -> uint32 (gosec)
        shouldRestart, _, _ := container.RestartManager().ShouldRestart(uint32(container.ExitCode()), container.HasBeenManuallyStopped, container.FinishedAt.Sub(container.StartedAt))
                                                                              ^
    container/view.go:401:25: G115: integer overflow conversion int -> uint16 (gosec)
                        PrivatePort: uint16(p),
                                           ^
    container/view.go:413:25: G115: integer overflow conversion int -> uint16 (gosec)
                        PrivatePort: uint16(p),
                                           ^
    container/view.go:414:25: G115: integer overflow conversion int -> uint16 (gosec)
                        PublicPort:  uint16(h),
                                           ^

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented:
Not sure if I've seen this before;

```
daemon.go:317: [dcae06e175a33] failed to start daemon: Error initializing network controller: list bridge addresses failed: interrupted system call
```

Failing here;

```
=== FAIL: amd64.integration.networking TestMACAddrOnRestart (0.51s)
    daemon.go:317: [dcae06e175a33] time="2024-08-28T14:24:47.663238628Z" level=debug msg="/usr/sbin/ip6tables, [--wait -D FORWARD -j DOCKER-USER]"
    daemon.go:317: [dcae06e175a33] time="2024-08-28T14:24:47.664238462Z" level=debug msg="/usr/sbin/ip6tables, [--wait -I FORWARD -j DOCKER-USER]"
    daemon.go:317: [dcae06e175a33] time="2024-08-28T14:24:47.666301302Z" level=debug msg="/usr/sbin/iptables, [--wait -t filter -n -L DOCKER-USER]"
    daemon.go:317: [dcae06e175a33] time="2024-08-28T14:24:47.667398408Z" level=debug msg="/usr/sbin/iptables, [--wait -t filter -C DOCKER-USER -j RETURN]"
    daemon.go:317: [dcae06e175a33] time="2024-08-28T14:24:47.668422989Z" level=debug msg="/usr/sbin/iptables, [--wait -t filter -C FORWARD -j DOCKER-USER]"
    daemon.go:317: [dcae06e175a33] time="2024-08-28T14:24:47.669597557Z" level=debug msg="/usr/sbin/iptables, [--wait -D FORWARD -j DOCKER-USER]"
    daemon.go:317: [dcae06e175a33] time="2024-08-28T14:24:47.670732765Z" level=debug msg="/usr/sbin/iptables, [--wait -I FORWARD -j DOCKER-USER]"
    daemon.go:317: [dcae06e175a33] time="2024-08-28T14:24:47.671802440Z" level=debug msg="/usr/sbin/ip6tables, [--wait -t filter -n -L DOCKER-USER]"
    daemon.go:317: [dcae06e175a33] time="2024-08-28T14:24:47.672795181Z" level=debug msg="/usr/sbin/ip6tables, [--wait -t filter -C DOCKER-USER -j RETURN]"
    daemon.go:317: [dcae06e175a33] time="2024-08-28T14:24:47.673773157Z" level=debug msg="/usr/sbin/ip6tables, [--wait -t filter -C FORWARD -j DOCKER-USER]"
    daemon.go:317: [dcae06e175a33] time="2024-08-28T14:24:47.674742955Z" level=debug msg="/usr/sbin/ip6tables, [--wait -D FORWARD -j DOCKER-USER]"
    daemon.go:317: [dcae06e175a33] time="2024-08-28T14:24:47.675738902Z" level=debug msg="/usr/sbin/ip6tables, [--wait -I FORWARD -j DOCKER-USER]"
    daemon.go:317: [dcae06e175a33] time="2024-08-28T14:24:47.678033911Z" level=debug msg="daemon configured with a 15 seconds minimum shutdown timeout"
    daemon.go:317: [dcae06e175a33] time="2024-08-28T14:24:47.678056563Z" level=debug msg="start clean shutdown of all containers with a 15 seconds timeout..."
    daemon.go:317: [dcae06e175a33] time="2024-08-28T14:24:47.678931938Z" level=debug msg="Unix socket /tmp/dxr/dcae06e175a33/libnetwork/ec8441010d15.sock was closed. The external key listener will stop."
    daemon.go:317: [dcae06e175a33] time="2024-08-28T14:24:47.679234395Z" level=info msg="stopping event stream following graceful shutdown" error="<nil>" module=libcontainerd namespace=dcae06e175a33
    daemon.go:317: [dcae06e175a33] time="2024-08-28T14:24:47.679266354Z" level=debug msg="Cleaning up old mountid : start."
    daemon.go:317: [dcae06e175a33] time="2024-08-28T14:24:47.679565555Z" level=debug msg="Cleaning up old mountid : done."
    daemon.go:317: [dcae06e175a33] time="2024-08-28T14:24:47.679833167Z" level=info msg="stopping event stream following graceful shutdown" error="context canceled" module=libcontainerd namespace=dcae06e175a33p
    daemon.go:317: [dcae06e175a33] failed to start daemon: Error initializing network controller: list bridge addresses failed: interrupted system call
    mac_addr_test.go:32: [dcae06e175a33] failed to start daemon with arguments [--data-root /go/src/github.com/docker/docker/bundles/test-integration/TestMACAddrOnRestart/dcae06e175a33/root --exec-root /tmp/dxr/dcae06e175a33 --pidfile /go/src/github.com/docker/docker/bundles/test-integration/TestMACAddrOnRestart/dcae06e175a33/docker.pid --userland-proxy=true --containerd-namespace dcae06e175a33 --containerd-plugins-namespace dcae06e175a33p --containerd /var/run/docker/containerd/containerd.sock --host unix:///tmp/docker-integration/dcae06e175a33.sock --debug --storage-driver overlay2] : [dcae06e175a33] daemon exited during startup: exit status 1
```

Some more similar ones;

```
=== Failed
=== FAIL: amd64.integration.container TestDaemonRestartKillContainers/live-restore=false/container_without_restart_policy/kill-daemon (1.60s)
    daemon.go:317: [dcc178fa40b1b] time="2024-08-28T14:24:36.887608345Z" level=debug msg="RequestAddress(LocalDefault/172.18.0.0/16, 172.18.0.2, map[])"
    daemon.go:317: [dcc178fa40b1b] time="2024-08-28T14:24:36.887618934Z" level=debug msg="Request address PoolID:172.18.0.0/16 Bits: 65536, Unselected: 65533, Sequence: (0xc0000000, 1)->(0x0, 2046)->(0x1, 1)->end Curr:0 Serial:false PrefAddress:172.18.0.2 "
    daemon.go:317: [dcc178fa40b1b] time="2024-08-28T14:24:36.888692653Z" level=info msg="Removing stale sandbox 1e4e7e7b41a07941c15323f5c141c2af360a57b518462e819815fe4e00795a69 (be81b6fa148a819da27edbdd6d992e20e8d8aa3d736cec0ff4227eafafaee8da)"
    daemon.go:317: [dcc178fa40b1b] time="2024-08-28T14:24:36.888761753Z" level=debug msg="Revoking external connectivity on endpoint laughing_rosalind (d9d1cb00f5362d67d884b6f6f694577e44c5a6e3c8cd3aa426c6f50f96990895)"
    daemon.go:317: [dcc178fa40b1b] time="2024-08-28T14:24:36.905353219Z" level=warning msg="Error (Unable to complete atomic operation, key modified) deleting object [endpoint 7a80c83ae4f80ccaed3ebb237ac286733e15c5e249225385d63fe5be243e7e9e d9d1cb00f5362d67d884b6f6f694577e44c5a6e3c8cd3aa426c6f50f96990895], retrying...."
    daemon.go:317: [dcc178fa40b1b] time="2024-08-28T14:24:36.905934107Z" level=debug msg="Releasing addresses for endpoint laughing_rosalind's interface on network bridge"
    daemon.go:317: [dcc178fa40b1b] time="2024-08-28T14:24:36.905969724Z" level=debug msg="ReleaseAddress(LocalDefault/172.18.0.0/16, 172.18.0.2)"
    daemon.go:317: [dcc178fa40b1b] time="2024-08-28T14:24:36.905992256Z" level=debug msg="Released address Address:172.18.0.2 Sequence:Bits: 65536, Unselected: 65533, Sequence: (0xc0000000, 1)->(0x0, 2046)->(0x1, 1)->end Curr:0"
    daemon.go:317: [dcc178fa40b1b] time="2024-08-28T14:24:36.908073596Z" level=debug msg="releasing IPv4 pools from network bridge (7a80c83ae4f80ccaed3ebb237ac286733e15c5e249225385d63fe5be243e7e9e)"
    daemon.go:317: [dcc178fa40b1b] time="2024-08-28T14:24:36.908093203Z" level=debug msg="ReleaseAddress(LocalDefault/172.18.0.0/16, 172.18.0.1)"
    daemon.go:317: [dcc178fa40b1b] time="2024-08-28T14:24:36.908109654Z" level=debug msg="Released address Address:172.18.0.1 Sequence:Bits: 65536, Unselected: 65534, Sequence: (0x80000000, 1)->(0x0, 2046)->(0x1, 1)->end Curr:0"
    daemon.go:317: [dcc178fa40b1b] time="2024-08-28T14:24:36.908121516Z" level=debug msg="ReleasePool(LocalDefault/172.18.0.0/16)"
    daemon.go:317: [dcc178fa40b1b] time="2024-08-28T14:24:36.908135302Z" level=debug msg="cleanupServiceDiscovery for network:7a80c83ae4f80ccaed3ebb237ac286733e15c5e249225385d63fe5be243e7e9e"
    daemon.go:317: [dcc178fa40b1b] time="2024-08-28T14:24:36.934007026Z" level=debug msg="daemon configured with a 15 seconds minimum shutdown timeout"
    daemon.go:317: [dcc178fa40b1b] time="2024-08-28T14:24:36.934025781Z" level=debug msg="start clean shutdown of all containers with a 15 seconds timeout..."
    daemon.go:317: [dcc178fa40b1b] time="2024-08-28T14:24:36.934597812Z" level=debug msg="Unix socket /tmp/dxr/dcc178fa40b1b/libnetwork/62a7498cb62f.sock was closed. The external key listener will stop."
    daemon.go:317: [dcc178fa40b1b] time="2024-08-28T14:24:36.935000646Z" level=debug msg="Cleaning up old mountid : start."
    daemon.go:317: [dcc178fa40b1b] time="2024-08-28T14:24:36.935069395Z" level=info msg="stopping event stream following graceful shutdown" error="<nil>" module=libcontainerd namespace=dcc178fa40b1b
    daemon.go:317: [dcc178fa40b1b] time="2024-08-28T14:24:36.935504832Z" level=debug msg="Cleaning up old mountid : done."
    daemon.go:317: [dcc178fa40b1b] failed to start daemon: Error initializing network controller: list bridge addresses failed: interrupted system call
    restart_test.go:120: [dcc178fa40b1b] failed to start daemon with arguments [--data-root /go/src/github.com/docker/docker/bundles/test-integration/TestDaemonRestartKillContainers/live-restore=false/container_without_restart_policy/kill-daemon/dcc178fa40b1b/root --exec-root /tmp/dxr/dcc178fa40b1b --pidfile /go/src/github.com/docker/docker/bundles/test-integration/TestDaemonRestartKillContainers/live-restore=false/container_without_restart_policy/kill-daemon/dcc178fa40b1b/docker.pid --userland-proxy=true --containerd-namespace dcc178fa40b1b --containerd-plugins-namespace dcc178fa40b1bp --containerd /var/run/docker/containerd/containerd.sock --host unix:///tmp/docker-integration/dcc178fa40b1b.sock --debug --storage-driver overlayfs --iptables=false --ip6tables=false] : [dcc178fa40b1b] daemon exited during startup: exit status 1
    panic.go:629: [dcc178fa40b1b] daemon is not started
    --- FAIL: TestDaemonRestartKillContainers/live-restore=false/container_without_restart_policy/kill-daemon (1.60s)
```

```
=== FAIL: amd64.integration.container TestDaemonRestartKillContainers/live-restore=true/container_without_restart_policy/stop-daemon (0.50s)
    daemon.go:317: [d9dc43e7ced30] time="2024-08-28T14:20:13.446040388Z" level=debug msg="Golang's threads limit set to 114930"
    daemon.go:317: [d9dc43e7ced30] time="2024-08-28T14:20:13.446692400Z" level=debug msg="metrics API listening on /tmp/dxr/d9dc43e7ced30/metrics.sock"
    daemon.go:317: [d9dc43e7ced30] time="2024-08-28T14:20:13.450513699Z" level=debug msg="Using default logging driver json-file"
    daemon.go:317: [d9dc43e7ced30] time="2024-08-28T14:20:13.450574784Z" level=debug msg="processing event stream" module=libcontainerd namespace=d9dc43e7ced30p
    daemon.go:317: [d9dc43e7ced30] time="2024-08-28T14:20:13.450816514Z" level=debug msg="No quota support for local volumes in /go/src/github.com/docker/docker/bundles/test-integration/TestDaemonRestartKillContainers/live-restore=true/container_without_restart_policy/stop-daemon/d9dc43e7ced30/root/volumes: Filesystem does not support, or has not enabled quotas"
    daemon.go:317: [d9dc43e7ced30] time="2024-08-28T14:20:13.453977595Z" level=warning msg="Enabling containerd snapshotter through the $TEST_INTEGRATION_USE_SNAPSHOTTER environment variable. This should only be used for testing."
    daemon.go:317: [d9dc43e7ced30] time="2024-08-28T14:20:13.453998274Z" level=info msg="Starting daemon with containerd snapshotter integration enabled"
    daemon.go:317: [d9dc43e7ced30] time="2024-08-28T14:20:13.454027288Z" level=info msg="Loading containers: start."
    daemon.go:317: [d9dc43e7ced30] time="2024-08-28T14:20:13.454073193Z" level=debug msg="Option DefaultDriver: bridge"
    daemon.go:317: [d9dc43e7ced30] time="2024-08-28T14:20:13.454083222Z" level=debug msg="Option DefaultNetwork: bridge"
    daemon.go:317: [d9dc43e7ced30] time="2024-08-28T14:20:13.454092139Z" level=debug msg="Network Control Plane MTU: 1500"
    daemon.go:317: [d9dc43e7ced30] time="2024-08-28T14:20:13.454205636Z" level=debug msg="processing event stream" module=libcontainerd namespace=d9dc43e7ced30
    daemon.go:317: [d9dc43e7ced30] time="2024-08-28T14:20:13.536241561Z" level=debug msg="daemon configured with a 15 seconds minimum shutdown timeout"
    daemon.go:317: [d9dc43e7ced30] time="2024-08-28T14:20:13.536274953Z" level=debug msg="start clean shutdown of all containers with a 15 seconds timeout..."
    daemon.go:317: [d9dc43e7ced30] time="2024-08-28T14:20:13.537380511Z" level=debug msg="Unix socket /tmp/dxr/d9dc43e7ced30/libnetwork/d5290b7af156.sock was closed. The external key listener will stop."
    daemon.go:317: [d9dc43e7ced30] time="2024-08-28T14:20:13.537703183Z" level=debug msg="Cleaning up old mountid : start."
    daemon.go:317: [d9dc43e7ced30] time="2024-08-28T14:20:13.537759111Z" level=info msg="stopping event stream following graceful shutdown" error="<nil>" module=libcontainerd namespace=d9dc43e7ced30
    daemon.go:317: [d9dc43e7ced30] time="2024-08-28T14:20:13.538100424Z" level=debug msg="Cleaning up old mountid : done."
    daemon.go:317: [d9dc43e7ced30] failed to start daemon: Error initializing network controller: list bridge addresses failed: interrupted system call
    daemon.go:317: [d9dc43e7ced30] time="2024-08-28T14:20:13.538447927Z" level=info msg="stopping event stream following graceful shutdown" error="context canceled" module=libcontainerd namespace=d9dc43e7ced30p
    restart_test.go:92: [d9dc43e7ced30] failed to start daemon with arguments [--data-root /go/src/github.com/docker/docker/bundles/test-integration/TestDaemonRestartKillContainers/live-restore=true/container_without_restart_policy/stop-daemon/d9dc43e7ced30/root --exec-root /tmp/dxr/d9dc43e7ced30 --pidfile /go/src/github.com/docker/docker/bundles/test-integration/TestDaemonRestartKillContainers/live-restore=true/container_without_restart_policy/stop-daemon/d9dc43e7ced30/docker.pid --userland-proxy=true --containerd-namespace d9dc43e7ced30 --containerd-plugins-namespace d9dc43e7ced30p --containerd /var/run/docker/containerd/containerd.sock --host unix:///tmp/docker-integration/d9dc43e7ced30.sock --debug --storage-driver overlayfs --iptables=false --ip6tables=false --live-restore] : [d9dc43e7ced30] daemon exited during startup: exit status 1
    --- FAIL: TestDaemonRestartKillContainers/live-restore=true/container_without_restart_policy/stop-daemon (0.50s)

=== FAIL: amd64.integration.container TestDaemonRestartKillContainers (0.00s)
```

corhere commented on hack/make/.binary:
```diff
@@ -78,6 +78,9 @@ source "${MAKEDIR}/.go-autogen"
 		GCFLAGS="all=-N -l"
 	fi
 
+	CGO_LDFLAGS=-latomic
+	export CGO_LDFLAGS
```
If this works, it may be worth factoring it out into a .go file so that it can be build-tagged to ARMv5, and so that plain old `go build` would also work without any special environment variables.

```go
//go:build arm.5

// #cgo LDFLAGS: -latomic
import "C"
```

thaJeztah replied:
Yeah, the amount of options we need is growing rapidly 😔

This was indeed just a quick push to see if it worked (almost forgot to add it again 😅)

corhere commented:
EINTR on netlink sockets is a new one. I suspect it has more to do with the netlink dependency bump you pulled in when rebasing than on the Go toolchain bump. I think the bug is here: https://github.com/vishvananda/netlink/blob/92645823f36c7ed03faf4baa566078d9d5e06fda/nl/nl_linux.go#L821-L824
It retries on `EWOULDBLOCK` (a.k.a. `EAGAIN`) but neglects to retry on `EINTR`.

thaJeztah commented:
Ah, nice find! Yeah, I wanted to have that update in master to get some burn-in time. Exactly for reasons like this, because there were a lot of changes since the last beta.

Worst case we could still revert that update (but not unlikely other dependencies will update it and force our hands)

thaJeztah commented:
Oh; looks like that code was added in v1.3.0, not v1.2.1 so we could revert that last bump only if needed 

- https://github.com/vishvananda/netlink/pull/793

thaJeztah commented:
Dropped the bump to go1.23, and moved that to https://github.com/moby/moby/pull/48404

thaJeztah commented on docker-bake.hcl:
```diff
@@ -81,7 +81,8 @@ group "default" {
   platforms = [
     "linux/amd64",
-    "linux/arm/v5",
+# FIXME(thaJeztah): temporarily disable arm/v5 as compiling is broken on go1.22; see https://github.com/moby/moby/pull/46982
+#     "linux/arm/v5",
     "linux/arm/v6",
```
We can use;

```
CGO_CFLAGS=-Wno-atomic-alignment CGO_LDFLAGS=-latomic
```

corhere commented on hack/make/.binary:
```diff
@@ -78,6 +78,17 @@ source "${MAKEDIR}/.go-autogen"
+		# cross-compiling for arm/v5 fails on go1.22; a fix is included for this
+		# in go1.23 (https://github.com/golang/go/issues/65290), but for go1.22
+		# we can set the correct option manually.
+		export CCGO_CFLAGS=-Wno-atomic-alignment
```
You misspelled `CGO_CFLAGS`, which is serendipitous: it doesn't clobber the `CGO_CFLAGS` variable set on line 25. But it means the flag doesn't actually take effect, either. Either use `+=` [or append the `-Wno-` flag  to line 25.](https://github.com/corhere/moby/commit/f94c328f2e711b92fe9065c3c73b833fec0f362a)

thaJeztah replied:
> You misspelled CGO_CFLAGS

DERP! My computer was jerky during the zoom call, and didn't notice 😂 


> [or append the -Wno- flag to line 25.](https://github.com/corhere/moby/commit/f94c328f2e711b92fe9065c3c73b833fec0f362a)

Oh! I like that one. I clearly didn't look at the file in a wider scope. Yes, that looks a lot cleaner as we already have the different variations there 👍

thaJeztah replied:
Oh! But I guess that branch is only for cross-compiling, so wouldn't apply to non-cross-compile 🤔 I can probably use the other approach; and I just spotted you did so as well in your branch in https://github.com/moby/moby/pull/48397

corhere committed:
hack/make/.binary: set CCGO_CFLAGS=-Wno-atomic-alignment for arm/v5

cross-compiling for arm/v5 fails on go1.22; a fix is included for this
in go1.23 (https://github.com/golang/go/issues/65290), but for go1.22
we can set the correct option manually.

    1.189 + go build -mod=vendor -modfile=vendor.mod -o /tmp/bundles/binary-daemon/dockerd -tags 'netgo osusergo static_build journald' -ldflags '-w -X "github.com/docker/docker/dockerversion.Version=dev" -X "github.com/docker/docker/dockerversion.GitCommit=HEAD" -X "github.com/docker/docker/dockerversion.BuildTime=2024-08-29T16:59:57.000000000+00:00" -X "github.com/docker/docker/dockerversion.PlatformName=" -X "github.com/docker/docker/dockerversion.ProductName=" -X "github.com/docker/docker/dockerversion.DefaultProductLicense=" -extldflags -static ' -gcflags= github.com/docker/docker/cmd/dockerd
    67.78 # runtime/cgo
    67.78 gcc_libinit.c:44:8: error: large atomic operation may incur significant performance penalty; the access size (4 bytes) exceeds the max lock-free size (0 bytes) [-Werror,-Watomic-alignment]
    67.78 gcc_libinit.c:47:6: error: large atomic operation may incur significant performance penalty; the access size (4 bytes) exceeds the max lock-free size (0 bytes) [-Werror,-Watomic-alignment]
    67.78 gcc_libinit.c:49:10: error: large atomic operation may incur significant performance penalty; the access size (4 bytes) exceeds the max lock-free size (0 bytes) [-Werror,-Watomic-alignment]
    67.78 gcc_libinit.c:69:9: error: large atomic operation may incur significant performance penalty; the access size (4 bytes) exceeds the max lock-free size (0 bytes) [-Werror,-Watomic-alignment]
    67.78 gcc_libinit.c:71:3: error: large atomic operation may incur significant performance penalty; the access size (4 bytes) exceeds the max lock-free size (0 bytes) [-Werror,-Watomic-alignment]
    78.20 + rm -f /go/src/github.com/docker/docker/go.mod

Co-authored-by: Sebastiaan van Stijn <github@gone.nl>
Signed-off-by: Cory Snider <csnider@mirantis.com>

corhere committed:
hack/make/.binary: set CGO_LDFLAGS=-latomic for arm/v5

cross-compiling for arm/v5 was failing;

    #56 84.12 /usr/bin/arm-linux-gnueabi-clang -marm -o $WORK/b001/exe/a.out -Wl,--export-dynamic-symbol=_cgo_panic -Wl,--export-dynamic-symbol=_cgo_topofstack -Wl,--export-dynamic-symbol=crosscall2 -Qunused-arguments -Wl,--compress-debug-sections=zlib /tmp/go-link-759578347/go.o /tmp/go-link-759578347/000000.o /tmp/go-link-759578347/000001.o /tmp/go-link-759578347/000002.o /tmp/go-link-759578347/000003.o /tmp/go-link-759578347/000004.o /tmp/go-link-759578347/000005.o /tmp/go-link-759578347/000006.o /tmp/go-link-759578347/000007.o /tmp/go-link-759578347/000008.o /tmp/go-link-759578347/000009.o /tmp/go-link-759578347/000010.o /tmp/go-link-759578347/000011.o /tmp/go-link-759578347/000012.o /tmp/go-link-759578347/000013.o /tmp/go-link-759578347/000014.o /tmp/go-link-759578347/000015.o /tmp/go-link-759578347/000016.o /tmp/go-link-759578347/000017.o /tmp/go-link-759578347/000018.o -O2 -g -O2 -g -O2 -g -lpthread -O2 -g -no-pie -static
    #56 84.12 ld.lld: error: undefined symbol: __atomic_load_4
    #56 84.12 >>> referenced by gcc_libinit.c
    #56 84.12 >>>               /tmp/go-link-759578347/000009.o:(_cgo_wait_runtime_init_done)
    #56 84.12 >>> referenced by gcc_libinit.c
    #56 84.12 >>>               /tmp/go-link-759578347/000009.o:(_cgo_wait_runtime_init_done)
    #56 84.12 >>> referenced by gcc_libinit.c
    #56 84.12 >>>               /tmp/go-link-759578347/000009.o:(_cgo_wait_runtime_init_done)
    #56 84.12 >>> referenced 2 more times
    #56 84.12
    #56 84.12 ld.lld: error: undefined symbol: __atomic_store_4
    #56 84.12 >>> referenced by gcc_libinit.c
    #56 84.12 >>>               /tmp/go-link-759578347/000009.o:(_cgo_wait_runtime_init_done)
    #56 84.12 >>> referenced by gcc_libinit.c
    #56 84.12 >>>               /tmp/go-link-759578347/000009.o:(x_cgo_notify_runtime_init_done)
    #56 84.12 >>> referenced by gcc_libinit.c
    #56 84.12 >>>               /tmp/go-link-759578347/000009.o:(x_cgo_set_context_function)
    #56 84.12 clang: error: linker command failed with exit code 1 (use -v to see invocation)

From discussion on GitHub;
https://github.com/moby/moby/pull/46982#issuecomment-2206992611

The arm/v5 build failure looks to be due to libatomic not being included
in the link. For reasons probably buried in mailing list archives,
[gcc](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=81358) and clang don't
bother to implicitly auto-link libatomic. This is not a big deal on many
modern platforms with atomic intrinsics as the compiler generates inline
instruction sequences, avoiding any libcalls into libatomic. ARMv5 is not
one of those platforms: all atomic operations require a libcall.

In theory, adding `CGO_LDFLAGS=-latomic` should fix arm/v5 builds.

While it could be argued that cgo should automatically link against
libatomic in the same way that it automatically links against libpthread,
the Go maintainers would have a valid counter-argument that it should be
the C toolchain's responsibility to link against libatomic automatically,
just like it does with libgcc or compiler-rt.

Co-authored-by: Sebastiaan van Stijn <github@gone.nl>
Signed-off-by: Cory Snider <csnider@mirantis.com>

thaJeztah commented:
Whoop! A couple of Flaky tests later, but it's green now 🎉

corhere commented on .github/workflows/.windows.yml:
```diff
@@ -91,7 +91,6 @@ jobs:
           & docker build `
             --build-arg WINDOWS_BASE_IMAGE `
             --build-arg WINDOWS_BASE_IMAGE_TAG `
-            --build-arg GO_VERSION `
```
Why are this line and the similar one below being removed? Won't it make the variable set in the `env:` block inert?

thaJeztah replied:
Hmm good question it's been a long time since I did this. I think this may have been to work around the "go versions don't have `.0` issue, and RC versions are not formatted as SemVer", and "actions/setup-go" only supporting SemVer, thus... could not use the same format.

I can add it back; that said, we don't run a matrix of versions in CI, and we always update both the Dockerfile and the GHA workflows, so in practice it's really redundant, so maybe removing it is the right thing to do?

corhere replied:
Maybe you're right; I just think we should give that particular change due consideration on its own merits rather than slipping it into this PR.

thaJeztah replied:
Yes, that makes sense; I removed it from this PR
